<a href="https://colab.research.google.com/github/shu-han-lei/Datasets_AI-final-project-2025Fall/blob/main/JAFS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import os
import glob
import pandas as pd

In [ ]:
from pathlib import Path


# Dataset

In [ ]:
# Folder Path

base = "/content/drive/MyDrive/AI_DATA"

folder_od  = f"{base}/OD_bronx_home_MSA"
folder_rac = f"{base}/RAC_BRONX"
folder_wac = f"{base}/WAC_MSA"

In [ ]:
files_od  = sorted(glob.glob(f"{folder_od}/*.parquet"))
files_rac = sorted(glob.glob(f"{folder_rac}/*.parquet"))
files_wac = sorted(glob.glob(f"{folder_wac}/*.parquet"))

In [ ]:
sample_od = pd.read_parquet(files_od[0])
sample_od.columns
sample_od


h_geocode        w_geocode  S000  SA01  SA02  SA03  SE01  SE02  \
0       360050002001001  360050002002004     1     0     1     0     0     1   
1       360050002002003  360050002002008     1     0     1     0     0     1   
2       360050041001000  360050002002008     1     1     0     0     0     1   
3       360050072001003  360050002002008     1     0     1     0     0     1   
4       360050084002003  360050002002008     1     0     1     0     1     0   
...                 ...              ...   ...   ...   ...   ...   ...   ...   
359844  360050093011002  421039501021015     1     1     0     0     0     1   
359845  360050068003002  421039502043043     1     0     1     0     1     0   
359846  360050328003002  421039508042024     1     0     1     0     0     0   
359847  360050350001003  421039508042024     1     0     1     0     1     0   
359848  360050409001004  421039508042024     1     0     1     0     1     0   

        SE03  SI01  SI02  SI03  year state  part work_fips is_CSA_work  \
0          0     0     0     1  2010    NY  main     36005           Y   
1          0     0     1     0  2010    NY  main     36005           Y   
2          0     0     1     0  2010    NY  main     36005           Y   
3          0     0     1     0  2010    NY  main     36005           Y   
4          0     0     1     0  2010    NY  main     36005           Y   
...      ...   ...   ...   ...   ...   ...   ...       ...         ...   
359844     0     0     0     1  2010    PA   aux     42103           Y   
359845     0     0     0     1  2010    PA   aux     42103           Y   
359846     1     0     0     1  2010    PA   aux     42103           Y   
359847     0     0     0     1  2010    PA   aux     42103           Y   
359848     0     0     0     1  2010    PA   aux     42103           Y   

            h_tract      w_tract  
0       36005000200  36005000200  
1       36005000200  36005000200  
2       36005004100  36005000200  
3       36005007200  36005000200  
4       36005008400  36005000200  
...             ...          ...  
359844  36005009301  42103950102  
359845  36005006800  42103950204  
359846  36005032800  42103950804  
359847  36005035000  42103950804  
359848  36005040900  42103950804  

[359849 rows x 19 columns]

In [ ]:
sample_wac = pd.read_parquet(files_wac[0])
sample_wac.columns
sample_wac


w_geocode  C000  CNS01  CNS02  CNS03  CNS04  CNS05  CNS06  \
0       360050002002004     1      0      0      0      0      0      0   
1       360050002002008     8      0      0      0      0      0      0   
2       360050002003000     2      0      0      0      0      0      0   
3       360050002003005     1      0      0      0      0      0      0   
4       360050002003007     2      0      0      0      0      0      0   
...                 ...   ...    ...    ...    ...    ...    ...    ...   
100807  421039509023003     2      0      0      0      0      0      0   
100808  421039509023004     7      0      0      0      0      0      0   
100809  421039509023014     3      0      0      0      1      0      0   
100810  421039509023028     3      0      0      0      0      0      0   
100811  421039509023032     4      0      0      0      0      0      0   

        CNS07  CNS08  ...  CD01  CD02  CD03  CD04  CS01  CS02  county_fips  \
0           0      0  ...     0     1     0     0     1     0        36005   
1           0      8  ...     1     2     2     2     7     1        36005   
2           0      0  ...     0     0     1     1     1     1        36005   
3           1      0  ...     0     0     0     0     0     1        36005   
4           0      0  ...     0     0     0     2     1     1        36005   
...       ...    ...  ...   ...   ...   ...   ...   ...   ...          ...   
100807      0      0  ...     0     2     0     0     1     1        42103   
100808      0      0  ...     0     1     3     1     5     2        42103   
100809      0      1  ...     0     2     1     0     2     1        42103   
100810      0      0  ...     1     0     2     0     1     2        42103   
100811      0      0  ...     0     3     1     0     2     2        42103   

        year  state      w_tract  
0       2010     NY  36005000200  
1       2010     NY  36005000200  
2       2010     NY  36005000200  
3       2010     NY  36005000200  
4       2010     NY  36005000200  
...      ...    ...          ...  
100807  2010     PA  42103950902  
100808  2010     PA  42103950902  
100809  2010     PA  42103950902  
100810  2010     PA  42103950902  
100811  2010     PA  42103950902  

[100812 rows x 38 columns]

In [ ]:
sample_rac = pd.read_parquet(files_rac[0])
sample_rac.columns
sample_rac

h_geocode  C000  CNS01  CNS02  CNS03  CNS04  CNS05  CNS06  CNS07  \
0     360050001001000   349      0      0      0     15     12     12     48   
1     360050001001001   326      1      0      0     17     18     14     46   
2     360050001001002    26      0      0      1      0      4      2      0   
3     360050001001003   591      1      0      4     24     25     24     82   
4     360050002001000   142      0      0      1      4      3     10     15   
...               ...   ...    ...    ...    ...    ...    ...    ...    ...   
4080  360050516014007    76      0      0      0      8      1      2      3   
4081  360050516014008    37      0      0      0      1      1      0      2   
4082  360050516014009    14      0      0      0      1      0      0      0   
4083  360050516014010    25      0      0      0      3      1      0      1   
4084  360050516014011     5      0      0      0      0      0      0      0   

      CNS08  ...  CD01  CD02  CD03  CD04  CS01  CS02  county_fips  year  \
0        36  ...    47    69    79    54   166   183        36005  2010   
1        28  ...    50    63    79    50   150   176        36005  2010   
2         2  ...     3     5     8     1    12    14        36005  2010   
3        54  ...    75   130   140    90   273   318        36005  2010   
4         3  ...    24    36    26    26    69    73        36005  2010   
...     ...  ...   ...   ...   ...   ...   ...   ...          ...   ...   
4080      5  ...     7    10    20    25    37    39        36005  2010   
4081      4  ...     0     7    12     8    22    15        36005  2010   
4082      0  ...     0     1     7     4     5     9        36005  2010   
4083      1  ...     2     3     3    12    15    10        36005  2010   
4084      1  ...     1     1     1     1     1     4        36005  2010   

      state      h_tract  
0        NY  36005000100  
1        NY  36005000100  
2        NY  36005000100  
3        NY  36005000100  
4        NY  36005000200  
...     ...          ...  
4080     NY  36005051601  
4081     NY  36005051601  
4082     NY  36005051601  
4083     NY  36005051601  
4084     NY  36005051601  

[4085 rows x 38 columns]

### MSA ANNUAL INCOME

In [ ]:
df_annual_income = pd.read_excel("/content/drive/MyDrive/AI_DATA/NAISC_median_hourly_wage.xlsx", sheet_name="NAICS ANNUAL INCOME")


In [ ]:
df_msa_income = df_annual_income[df_annual_income["MSA Code"] == 35620].copy()
df_msa_income.head()


CNS01         CNS02         CNS03         CNS04         CNS05  \
13  25679.696443  59490.384190  79995.813439  52427.098814  47800.373123   
14  25639.376973  60701.772833  82451.415265  53579.723674  48703.859250   
15  25640.122762  62381.873146  83794.103325  54171.578517  49043.553964   
16  25756.282828  63011.393939  86719.191919  54926.707071  49491.393939   
17  26336.706468  63968.330348  89001.777114  55279.932338  49795.381095   

           CNS06         CNS07         CNS08         CNS09         CNS10  ...  \
13  50072.670356  29293.470356  51496.278261  66170.390514  56533.660079  ...   
14  50977.436481  29803.036481  52894.911255  68289.494437  58482.980595  ...   
15  51443.905882  29840.738619  52998.679284  69746.589258  59408.709974  ...   
16  52113.454545  30181.010101  52687.030303  71341.898990  60908.282828  ...   
17  52374.749254  30382.241791  52890.622886  72792.484577  62067.743284  ...   

           CNS14         CNS15         CNS16         CNS17         CNS18  \
13  34686.754150  56752.676680  44706.763636  30717.078261  24776.252964   
14  35062.395860  57606.420699  45827.647089  30926.128849  25036.742044   
15  34750.549361  57553.892583  46206.774425  31068.191304  25040.034783   
16  35397.818182  57958.464646  46869.333333  31382.787879  25182.707071   
17  35703.885572  58402.325373  47188.861692  31712.652736  25250.656716   

           CNS19         CNS20  Year                             Area Title  \
13  36247.247431  60147.433992  2010  New York-Newark-Jersey City, NY-NJ-PA   
14  36651.160673  61277.015265  2011  New York-Newark-Jersey City, NY-NJ-PA   
15  36932.687468  61590.848082  2012  New York-Newark-Jersey City, NY-NJ-PA   
16  37391.676768  62465.131313  2013  New York-Newark-Jersey City, NY-NJ-PA   
17  37984.590050  63778.271642  2014  New York-Newark-Jersey City, NY-NJ-PA   

    MSA Code  
13   35620.0  
14   35620.0  
15   35620.0  
16   35620.0  
17   35620.0  

[5 rows x 23 columns]

In [ ]:
# Set Year as the index (to make it easier to look up salaries by year).
df_msa_income = df_msa_income.set_index("Year")

In [ ]:
df_msa_income.loc[2010]["CNS15"]

np.float64(56752.6766798419)

In [ ]:
# exclude CNS01 CNS02 CNS20
exclude = ["CNS01", "CNS02", "CNS20"]

CNS_cols = [
    c for c in df_msa_income.columns
    if c.startswith("CNS") and c not in exclude
]

print("Using CNS columns:", CNS_cols)


Using CNS columns: ['CNS03', 'CNS04', 'CNS05', 'CNS06', 'CNS07', 'CNS08', 'CNS09', 'CNS10', 'CNS11', 'CNS12', 'CNS13', 'CNS14', 'CNS15', 'CNS16', 'CNS17', 'CNS18', 'CNS19']


In [ ]:
# Establish an annual salary lookup (more suitable for JAFS pipelines)
wage_lookup = df_msa_income[CNS_cols].to_dict(orient="index")
wage_lookup[2010]["CNS15"]

56752.6766798419

选 NYC–MSA	df_annual_income[df["MSA Code"]==35620]

按年索引	df_nyc_income.set_index("Year")

每年 Caster job wage	df_nyc_income.loc[year]["CNSxx"]

Lookup dict（最适合 pipeline）	wage_lookup = df_nyc_income[CNS_cols].to_dict("index")

In [ ]:
df_HUD = pd.read_excel("/content/drive/MyDrive/AI_DATA/HOME Income Limits NYC Metro.xlsx")
df_HUD

Year  30% AMI  50% AMI  60% AMI  80% AMI  100% AMI  120% AMI  150% AMI
0   2010    16650    27750    33300    44350     55500     66600     83250
1   2011    17200    28650    34380    45850     57300     68760     85950
2   2012    17450    29050    34860    46500     58100     69720     87150
3   2013    18050    30100    36120    48100     60200     72240     90300
4   2014    17650    29400    35280    47000     58800     70560     88200
5   2015    18150    30250    36300    48350     60500     72600     90750
6   2016    19050    31750    38100    50750     63500     76200     95250
7   2017    20050    33400    40080    53450     66800     80160    100200
8   2018    21950    36550    43860    58450     73100     87720    109650
9   2019    22400    37350    44820    59750     74700     89640    112050
10  2020    23900    39800    47760    63700     79600     95520    119400
11  2021    25100    41800    50160    66850     83600    100320    125400
12  2022    28000    46700    56040    74800     93400    112080    140100

LOW-INCOME (0–80% AMI)
    - 30% AMI
    - 50% AMI
    - 60% AMI
    - 80% AMI

MEDIAN-INCOME (80–120% AMI)

HIGH-INCOME (120–150% AMI)


In [ ]:
def get_income_category(wage, year, df_HUD):
    thr30 = df_HUD.loc[df_HUD["Year"] == year, "30% AMI"].values[0]
    thr50 = df_HUD.loc[df_HUD["Year"] == year, "50% AMI"].values[0]
    thr60 = df_HUD.loc[df_HUD["Year"] == year, "60% AMI"].values[0]
    thr80 = df_HUD.loc[df_HUD["Year"] == year, "80% AMI"].values[0]
    thr100 = df_HUD.loc[df_HUD["Year"] == year, "100% AMI"].values[0]
    thr120 = df_HUD.loc[df_HUD["Year"] == year, "120% AMI"].values[0]

    # ---- LOW-INCOME ----
    if wage < thr30:
        return "Low-30"
    elif wage < thr50:
        return "Low-50"
    elif wage < thr60:
        return "Low-60"
    elif wage < thr80:
        return "Low-80"

    # ---- MEDIAN-INCOME ----
    elif wage < thr120:
        return "Median"

    # ---- HIGH-INCOME ----
    else:
        return "High"



In [ ]:
# 把 HUD 表按 Year 设为 index，方便取某一年的阈值
df_HUD = df_HUD.set_index("Year")

# OES 工资表你已经 set_index("Year") 了：
# df_msa_income = df_msa_income.set_index("Year")

# 从 OES 表做一个 lookup: wage_lookup[2015]["CNS15"] -> 年收入
wage_lookup = df_msa_income[CNS_cols].to_dict(orient="index")


### Yearly JAFS Pipeline -  OD × WAC Detailed Merged Dataset (with w_tract)

In [ ]:
base = "/content/drive/MyDrive/AI_DATA"
jafs_dir = Path(f"{base}/JAFS_BRONX_wage_HUD")
jafs_dir.mkdir(exist_ok=True)

In [ ]:
all_year_dfs = []
years_sorted = sorted(data_od.keys())   # ['2010', ..., '2022']

for year in years_sorted:
    print(f"\n=== Year {year} ===")
    year_int = int(year)

    # ---------- 1) 取当年 OD / WAC / RAC ----------
    df_od  = data_od[year]      # Bronx home OD
    df_wac = data_wac[year]     # MSA WAC
    df_rac = data_rac[year]     # Bronx-only RAC

    # WAC 只取需要列
    wac_use_cols = ["w_tract", "year", "C000"] + CNS_cols
    df_wac_use = df_wac[wac_use_cols].copy()

    # ---------- 2) OD + WAC merge ----------
    df = df_od.merge(df_wac_use, on=["w_tract", "year"], how="left")
    df[["C000"] + CNS_cols] = df[["C000"] + CNS_cols].fillna(0)

    # ---------- 3) 当年 HUD 阈值 ----------
    limits  = df_HUD.loc[year_int]
    thr30   = limits["30% AMI"]
    thr50   = limits["50% AMI"]
    thr60   = limits["60% AMI"]
    thr80   = limits["80% AMI"]
    thr120  = limits["120% AMI"]    # 用于 MEDIAN / HIGH 分界

    # ---------- 4) 当年各行业年收入（OES） ----------
    wages = wage_lookup[year_int]   # dict: CNSxx -> annual wage

    # 大类：LOW / MEDIAN / HIGH
    low_cns     = [c for c in CNS_cols if wages[c] <= thr80]
    median_cns  = [c for c in CNS_cols if thr80 < wages[c] <= thr120]
    high_cns    = [c for c in CNS_cols if wages[c] > thr120]

    # LOW 里面的细分：30 / 50 / 60 / 80
    low30_cns = [c for c in CNS_cols if wages[c] <= thr30]
    low50_cns = [c for c in CNS_cols if thr30 < wages[c] <= thr50]
    low60_cns = [c for c in CNS_cols if thr50 < wages[c] <= thr60]
    low80_cns = [c for c in CNS_cols if thr60 < wages[c] <= thr80]

    print("  low_cns   :", low_cns)
    print("  median_cns:", median_cns)
    print("  high_cns  :", high_cns)

    # ---------- 5) 计算各类可达岗位（flow S000 × jobs） ----------
    df["access_jobs_total"] = df["S000"] * df["C000"]

    # 大类
    df["access_low_total"]    = df["S000"] * df[low_cns].sum(axis=1)
    df["access_median_total"] = df["S000"] * df[median_cns].sum(axis=1)
    df["access_high_total"]   = df["S000"] * df[high_cns].sum(axis=1)

    # LOW 子类（可选，如果你只想大类，可以跳过）
    df["access_low30"] = df["S000"] * (df[low30_cns].sum(axis=1) if low30_cns else 0)
    df["access_low50"] = df["S000"] * (df[low50_cns].sum(axis=1) if low50_cns else 0)
    df["access_low60"] = df["S000"] * (df[low60_cns].sum(axis=1) if low60_cns else 0)
    df["access_low80"] = df["S000"] * (df[low80_cns].sum(axis=1) if low80_cns else 0)

    # ---------- 6) 聚合到 Bronx home tract ----------
    agg_cols = [
        "access_jobs_total",
        "access_low_total", "access_median_total", "access_high_total",
        "access_low30", "access_low50", "access_low60", "access_low80"
    ]

    df_jafs = (
        df.groupby(["h_tract", "year"])[agg_cols]
          .sum()
          .reset_index()
    )

    # ---------- 7) 合并 RAC（人口作为分母） ----------
    df_rac_use = (
        df_rac[["h_tract", "C000"]]
        .rename(columns={"C000": "resident_pop"})
        .drop_duplicates("h_tract")
    )

    df_jafs = df_jafs.merge(df_rac_use, on="h_tract", how="left")
    denom = df_jafs["resident_pop"].replace(0, pd.NA)

    # per capita 指标
    df_jafs["JAFS_total_pc"]   = df_jafs["access_jobs_total"]   / denom
    df_jafs["JAFS_low_pc"]     = df_jafs["access_low_total"]    / denom
    df_jafs["JAFS_median_pc"]  = df_jafs["access_median_total"] / denom
    df_jafs["JAFS_high_pc"]    = df_jafs["access_high_total"]   / denom

    df_jafs["JAFS_low30_pc"]   = df_jafs["access_low30"] / denom
    df_jafs["JAFS_low50_pc"]   = df_jafs["access_low50"] / denom
    df_jafs["JAFS_low60_pc"]   = df_jafs["access_low60"] / denom
    df_jafs["JAFS_low80_pc"]   = df_jafs["access_low80"] / denom

    # ---------- 8) 保存该年份结果 ----------
    out_path = jafs_dir / f"JAFS_bronx_HUDwage2_{year}.parquet"
    df_jafs.to_parquet(out_path, index=False)
    print(f"  Saved {out_path.name}, rows={len(df_jafs)}")

    all_year_dfs.append(df_jafs)

# ---------- 9) 拼成 panel ----------
jafs_panel = pd.concat(all_year_dfs, ignore_index=True)
panel_path = jafs_dir / "JAFS_bronx_HUDwage_panel_2010_2022.parquet"
jafs_panel.to_parquet(panel_path, index=False)
print(f"\nPanel dataset saved: {panel_path}")



=== Year 2010 ===
  low_cns   : ['CNS07', 'CNS11', 'CNS14', 'CNS17', 'CNS18', 'CNS19']
  median_cns: ['CNS04', 'CNS05', 'CNS06', 'CNS08', 'CNS09', 'CNS10', 'CNS15', 'CNS16']
  high_cns  : ['CNS03', 'CNS12', 'CNS13']
  Saved JAFS_bronx_HUDwage2_2010.parquet, rows=358

=== Year 2011 ===
  low_cns   : ['CNS07', 'CNS11', 'CNS14', 'CNS16', 'CNS17', 'CNS18', 'CNS19']
  median_cns: ['CNS04', 'CNS05', 'CNS06', 'CNS08', 'CNS09', 'CNS10', 'CNS15']
  high_cns  : ['CNS03', 'CNS12', 'CNS13']
  Saved JAFS_bronx_HUDwage2_2011.parquet, rows=354

=== Year 2012 ===
  low_cns   : ['CNS07', 'CNS11', 'CNS14', 'CNS16', 'CNS17', 'CNS18', 'CNS19']
  median_cns: ['CNS04', 'CNS05', 'CNS06', 'CNS08', 'CNS10', 'CNS15']
  high_cns  : ['CNS03', 'CNS09', 'CNS12', 'CNS13']
  Saved JAFS_bronx_HUDwage2_2012.parquet, rows=353

=== Year 2013 ===
  low_cns   : ['CNS07', 'CNS11', 'CNS14', 'CNS16', 'CNS17', 'CNS18', 'CNS19']
  median_cns: ['CNS04', 'CNS05', 'CNS06', 'CNS08', 'CNS09', 'CNS10', 'CNS15']
  high_cns  : ['CNS03

In [ ]:
df_JAFS = pd.read_parquet("/content/drive/MyDrive/AI_DATA/JAFS_BRONX_wage_HUD/JAFS_bronx_HUDwage_panel_2010_2022.parquet")


h_tract  year  access_jobs_total  access_low_total  \
0  36005000100  2010           39039455           1752466   
1  36005000200  2010           27237182           3959411   
2  36005000400  2010           31967034           4772822   
3  36005001600  2010           39808325           5697632   
4  36005001901  2010            1985105            406442   

   access_median_total  access_high_total  access_low30  access_low50  \
0             35493656             906596             0        335374   
1             18355470            2384769             0        828657   
2             21579992            2930698             0       1007148   
3             26905254            3489877             0       1235032   
4               910977             250150             0         97799   

   access_low60  access_low80  resident_pop  JAFS_total_pc    JAFS_low_pc  \
0        574332        842760           349  111860.902579    5021.392550   
1       1032952       2097802           142  191811.140845   27883.176056   
2       1265549       2500125           281  113761.686833   16985.131673   
3       1521902       2940698          1042   38203.766795    5467.976967   
4        117755        190888             3  661701.666667  135480.666667   

   JAFS_median_pc  JAFS_high_pc  JAFS_low30_pc  JAFS_low50_pc  JAFS_low60_pc  \
0   101701.020057   2597.696275            0.0     960.957020    1645.650430   
1   129263.873239  16794.147887            0.0    5835.612676    7274.309859   
2    76797.124555  10429.530249            0.0    3584.156584    4503.733096   
3    25820.781190   3349.210173            0.0    1185.251440    1460.558541   
4   303659.000000  83383.333333            0.0   32599.666667   39251.666667   

   JAFS_low80_pc  
0    2414.785100  
1   14773.253521  
2    8897.241993  
3    2822.166987  
4   63629.333333

In [ ]:
df_JAFS

h_tract  year  access_jobs_total  access_low_total  \
0     36005000100  2010           39039455           1752466   
1     36005000200  2010           27237182           3959411   
2     36005000400  2010           31967034           4772822   
3     36005001600  2010           39808325           5697632   
4     36005001901  2010            1985105            406442   
...           ...   ...                ...               ...   
4621  36005048401  2022           13193852           8317865   
4622  36005048402  2022             697069            440698   
4623  36005050400  2022            1065668            668453   
4624  36005051601  2022           17922361          10404970   
4625  36005051602  2022            1023990            640188   

      access_median_total  access_high_total  access_low30  access_low50  \
0                35493656             906596             0        335374   
1                18355470            2384769             0        828657   
2                21579992            2930698             0       1007148   
3                26905254            3489877             0       1235032   
4                  910977             250150             0         97799   
...                   ...                ...           ...           ...   
4621              4083509              49112             0       1299162   
4622               216419                959             0         71490   
4623               371964                323             0        131238   
4624              6408845              91992             0       1944086   
4625               345693                834             0        130593   

      access_low60  access_low80  resident_pop  JAFS_total_pc    JAFS_low_pc  \
0           574332        842760           349  111860.902579    5021.392550   
1          1032952       2097802           142  191811.140845   27883.176056   
2          1265549       2500125           281  113761.686833   16985.131673   
3          1521902       2940698          1042   38203.766795    5467.976967   
4           117755        190888             3  661701.666667  135480.666667   
...            ...           ...           ...            ...            ...   
4621       1136953       5881750            40  329846.300000  207946.625000   
4622         53054        316154           131    5321.137405    3364.106870   
4623        119487        417728           132    8073.242424    5064.037879   
4624       1670065       6790819            50  358447.220000  208099.400000   
4625        124411        385184            85   12046.941176    7531.623529   

      JAFS_median_pc  JAFS_high_pc  JAFS_low30_pc  JAFS_low50_pc  \
0      101701.020057   2597.696275            0.0     960.957020   
1      129263.873239  16794.147887            0.0    5835.612676   
2       76797.124555  10429.530249            0.0    3584.156584   
3       25820.781190   3349.210173            0.0    1185.251440   
4      303659.000000  83383.333333            0.0   32599.666667   
...              ...           ...            ...            ...   
4621   102087.725000   1227.800000            0.0   32479.050000   
4622     1652.053435      7.320611            0.0     545.725191   
4623     2817.909091      2.446970            0.0     994.227273   
4624   128176.900000   1839.840000            0.0   38881.720000   
4625     4066.976471      9.811765            0.0    1536.388235   

      JAFS_low60_pc  JAFS_low80_pc  
0       1645.650430    2414.785100  
1       7274.309859   14773.253521  
2       4503.733096    8897.241993  
3       1460.558541    2822.166987  
4      39251.666667   63629.333333  
...             ...            ...  
4621   28423.825000  147043.750000  
4622     404.992366    2413.389313  
4623     905.204545    3164.606061  
4624   33401.300000  135816.380000  
4625    1463.658824    4531.576471  

[4626 rows x 19 columns]

In [ ]:
df_JAFS_2022 = pd.read_parquet("/content/drive/MyDrive/AI_DATA/JAFS_BRONX_wage_HUD/JAFS_bronx_HUDwage_2022.parquet")


In [ ]:
df_JAFS_2022

h_tract  year  access_jobs_total  access_low_total  \
0    36005000100  2022            3931761           2966555   
1    36005000200  2022           16710464          10133205   
2    36005000400  2022           24215019          14420841   
3    36005001600  2022           24170932          14548245   
4    36005001901  2022           14545039           7891939   
..           ...   ...                ...               ...   
356  36005048401  2022           13193852           8317865   
357  36005048402  2022             697069            440698   
358  36005050400  2022            1065668            668453   
359  36005051601  2022           17922361          10404970   
360  36005051602  2022            1023990            640188   

     access_median_total  access_high_total  access_low30  access_low50  \
0                 606049              36106             0        371928   
1                5463641             118558             0       1662891   
2                8342360             135192             0       2382016   
3                8240186             110803             0       2407632   
4                5949956              69273             0       1592733   
..                   ...                ...           ...           ...   
356              4083509              49112             0       1299162   
357               216419                959             0         71490   
358               371964                323             0        131238   
359              6408845              91992             0       1944086   
360               345693                834             0        130593   

     access_low60  access_low80  resident_pop  JAFS_total_pc    JAFS_low_pc  \
0          239823       2354804            20  196588.050000  148327.750000   
1         1619323       6850991            90  185671.822222  112591.166667   
2         2292505       9746320           186  130188.274194   77531.403226   
3         2376675       9763938           864   27975.615741   16838.246528   
4         1606734       4692472            41  354757.048780  192486.317073   
..            ...           ...           ...            ...            ...   
356       1136953       5881750            40  329846.300000  207946.625000   
357         53054        316154           131    5321.137405    3364.106870   
358        119487        417728           132    8073.242424    5064.037879   
359       1670065       6790819            50  358447.220000  208099.400000   
360        124411        385184            85   12046.941176    7531.623529   

     JAFS_median_pc  JAFS_high_pc  JAFS_low30_pc  JAFS_low50_pc  \
0      30302.450000   1805.300000            0.0   18596.400000   
1      60707.122222   1317.311111            0.0   18476.566667   
2      44851.397849    726.838710            0.0   12806.537634   
3       9537.252315    128.244213            0.0    2786.611111   
4     145120.878049   1689.585366            0.0   38847.146341   
..              ...           ...            ...            ...   
356   102087.725000   1227.800000            0.0   32479.050000   
357     1652.053435      7.320611            0.0     545.725191   
358     2817.909091      2.446970            0.0     994.227273   
359   128176.900000   1839.840000            0.0   38881.720000   
360     4066.976471      9.811765            0.0    1536.388235   

     JAFS_low60_pc  JAFS_low80_pc  
0     11991.150000  117740.200000  
1     17992.477778   76122.122222  
2     12325.295699   52399.569892  
3      2750.781250   11300.854167  
4     39188.634146  114450.536585  
..             ...            ...  
356   28423.825000  147043.750000  
357     404.992366    2413.389313  
358     905.204545    3164.606061  
359   33401.300000  135816.380000  
360    1463.658824    4531.576471  

[361 rows x 19 columns]

In [ ]:
df_JAFS_2022.columns

Index(['h_tract', 'year', 'access_jobs_total', 'access_low_total',
       'access_median_total', 'access_high_total', 'access_low30',
       'access_low50', 'access_low60', 'access_low80', 'resident_pop',
       'JAFS_total_pc', 'JAFS_low_pc', 'JAFS_median_pc', 'JAFS_high_pc',
       'JAFS_low30_pc', 'JAFS_low50_pc', 'JAFS_low60_pc', 'JAFS_low80_pc'],
      dtype='object')

In [ ]:
# ==== Macro Sector ====
macro_sector_map = {
    "Office": [
        "CNS09", "CNS10", "CNS11", "CNS12", "CNS13"
    ],
    "Industrial": [
        "CNS03", "CNS04", "CNS05", "CNS06", "CNS08"
    ],
    "Institutional": [
        "CNS15", "CNS16"
    ],
    "LocalServices": [
        "CNS07", "CNS14", "CNS17", "CNS18", "CNS19"
    ]
}


In [ ]:
all_year_dfs = []
years_sorted = sorted(data_od.keys())   # ['2010', ..., '2022']

for year in years_sorted:
    print(f"\n=== Year {year} ===")
    year_int = int(year)

    # ---------- 1) 取当年 OD / WAC / RAC ----------
    df_od  = data_od[year]      # Bronx home OD
    df_wac = data_wac[year]     # MSA WAC
    df_rac = data_rac[year]     # Bronx-only RAC

    # WAC 只取需要列
    wac_use_cols = ["w_tract", "year", "C000"] + CNS_cols
    df_wac_use = df_wac[wac_use_cols].copy()

    # ---------- 2) OD + WAC merge ----------
    df = df_od.merge(df_wac_use, on=["w_tract", "year"], how="left")
    df[["C000"] + CNS_cols] = df[["C000"] + CNS_cols].fillna(0)

    # ---------- 3) 当年 HUD 阈值 ----------
    limits  = df_HUD.loc[year_int]
    thr30   = limits["30% AMI"]
    thr50   = limits["50% AMI"]
    thr60   = limits["60% AMI"]
    thr80   = limits["80% AMI"]
    thr120  = limits["120% AMI"]    # 用于 MEDIAN / HIGH 分界

    # ---------- 4) 当年各行业年收入（OES） ----------
    wages = wage_lookup[year_int]   # dict: CNSxx -> annual wage

    # 大类：LOW / MEDIAN / HIGH
    low_cns     = [c for c in CNS_cols if wages[c] <= thr80]
    median_cns  = [c for c in CNS_cols if thr80 < wages[c] <= thr120]
    high_cns    = [c for c in CNS_cols if wages[c] > thr120]

    # LOW 里面的细分：30 / 50 / 60 / 80
    low30_cns = [c for c in CNS_cols if wages[c] <= thr30]
    low50_cns = [c for c in CNS_cols if thr30 < wages[c] <= thr50]
    low60_cns = [c for c in CNS_cols if thr50 < wages[c] <= thr60]
    low80_cns = [c for c in CNS_cols if thr60 < wages[c] <= thr80]

    print("  low_cns   :", low_cns)
    print("  median_cns:", median_cns)
    print("  high_cns  :", high_cns)

    # ---------- 5) 计算各类可达岗位（flow S000 × jobs） ----------
    df["access_jobs_total"] = df["S000"] * df["C000"]

    df["access_low_total"]    = df["S000"] * (df[low_cns].sum(axis=1) if low_cns else 0)
    df["access_median_total"] = df["S000"] * (df[median_cns].sum(axis=1) if median_cns else 0)
    df["access_high_total"]   = df["S000"] * (df[high_cns].sum(axis=1) if high_cns else 0)

    df["access_low30"] = df["S000"] * (df[low30_cns].sum(axis=1) if low30_cns else 0)
    df["access_low50"] = df["S000"] * (df[low50_cns].sum(axis=1) if low50_cns else 0)
    df["access_low60"] = df["S000"] * (df[low60_cns].sum(axis=1) if low60_cns else 0)
    df["access_low80"] = df["S000"] * (df[low80_cns].sum(axis=1) if low80_cns else 0)

    # ---------- 5c) Macro Sector job totals & accessibility ----------
    for sector, cns_list in macro_sector_map.items():
        valid_cns = [c for c in cns_list if c in df.columns]
        if valid_cns:
            df[f"jobs_{sector}"]    = df[valid_cns].sum(axis=1)
            df[f"access_{sector}"]  = df["S000"] * df[f"jobs_{sector}"]
        else:
            df[f"jobs_{sector}"]    = 0
            df[f"access_{sector}"]  = 0

    # ---------- 5b) 保存 OD × WAC 详细数据（含 w_tract + Macro Sector） ----------
    detail_cols = (
        ["h_tract", "w_tract", "year", "S000"] +
        ["C000"] + CNS_cols +
        ["access_jobs_total", "access_low_total",
         "access_median_total", "access_high_total"] +
        [f"jobs_{s}" for s in macro_sector_map.keys()] +
        [f"access_{s}" for s in macro_sector_map.keys()]
    )

    df_detail = df[detail_cols].copy()
    detail_path = jafs_dir / f"OD_WAC_detail_{year}.parquet"
    df_detail.to_parquet(detail_path, index=False)
    print(f"  Saved detailed OD-WAC dataset: {detail_path.name}, rows={len(df_detail)}")

    # ---------- 6) 聚合到 Bronx home tract ----------
    agg_cols = [
        "access_jobs_total",
        "access_low_total", "access_median_total", "access_high_total",
        "access_low30", "access_low50", "access_low60", "access_low80"
    ] + [f"access_{s}" for s in macro_sector_map.keys()]

    df_jafs = (
        df.groupby(["h_tract", "year"])[agg_cols]
          .sum()
          .reset_index()
    )

    # ---------- 7) 合并 RAC（人口作为分母） ----------
    df_rac_use = (
        df_rac[["h_tract", "C000"]]
        .rename(columns={"C000": "resident_pop"})
        .drop_duplicates("h_tract")
    )

    df_jafs = df_jafs.merge(df_rac_use, on="h_tract", how="left")
    denom = df_jafs["resident_pop"].replace(0, pd.NA)

    # per capita 指标（工资桶）
    df_jafs["JAFS_total_pc"]   = df_jafs["access_jobs_total"]   / denom
    df_jafs["JAFS_low_pc"]     = df_jafs["access_low_total"]    / denom
    df_jafs["JAFS_median_pc"]  = df_jafs["access_median_total"] / denom
    df_jafs["JAFS_high_pc"]    = df_jafs["access_high_total"]   / denom

    df_jafs["JAFS_low30_pc"]   = df_jafs["access_low30"] / denom
    df_jafs["JAFS_low50_pc"]   = df_jafs["access_low50"] / denom
    df_jafs["JAFS_low60_pc"]   = df_jafs["access_low60"] / denom
    df_jafs["JAFS_low80_pc"]   = df_jafs["access_low80"] / denom

    # per capita 指标（Macro Sectors）
    for sector in macro_sector_map.keys():
        df_jafs[f"JAFS_{sector}_pc"] = df_jafs[f"access_{sector}"] / denom

    # ---------- 8) 保存该年份结果 ----------
    out_path = jafs_dir / f"JAFS_bronx_HUDwage_macro_{year}.parquet"
    df_jafs.to_parquet(out_path, index=False)
    print(f"  Saved {out_path.name}, rows={len(df_jafs)}")

    all_year_dfs.append(df_jafs)

# ---------- 9) 拼成 panel ----------
jafs_panel = pd.concat(all_year_dfs, ignore_index=True)
panel_path = jafs_dir / "JAFS_bronx_HUDwage_macro_panel_2010_2022.parquet"
jafs_panel.to_parquet(panel_path, index=False)
print(f"\nPanel dataset saved: {panel_path}")



=== Year 2010 ===
  low_cns   : ['CNS07', 'CNS11', 'CNS14', 'CNS17', 'CNS18', 'CNS19']
  median_cns: ['CNS04', 'CNS05', 'CNS06', 'CNS08', 'CNS09', 'CNS10', 'CNS15', 'CNS16']
  high_cns  : ['CNS03', 'CNS12', 'CNS13']
  Saved detailed OD-WAC dataset: OD_WAC_detail_2010.parquet, rows=6505737
  Saved JAFS_bronx_HUDwage_macro_2010.parquet, rows=358

=== Year 2011 ===
  low_cns   : ['CNS07', 'CNS11', 'CNS14', 'CNS16', 'CNS17', 'CNS18', 'CNS19']
  median_cns: ['CNS04', 'CNS05', 'CNS06', 'CNS08', 'CNS09', 'CNS10', 'CNS15']
  high_cns  : ['CNS03', 'CNS12', 'CNS13']
  Saved detailed OD-WAC dataset: OD_WAC_detail_2011.parquet, rows=6113677
  Saved JAFS_bronx_HUDwage_macro_2011.parquet, rows=354

=== Year 2012 ===
  low_cns   : ['CNS07', 'CNS11', 'CNS14', 'CNS16', 'CNS17', 'CNS18', 'CNS19']
  median_cns: ['CNS04', 'CNS05', 'CNS06', 'CNS08', 'CNS10', 'CNS15']
  high_cns  : ['CNS03', 'CNS09', 'CNS12', 'CNS13']
  Saved detailed OD-WAC dataset: OD_WAC_detail_2012.parquet, rows=6443372
  Saved JAFS_br

In [ ]:
df_JAFS_2022 = pd.read_parquet("/content/drive/MyDrive/AI_DATA/JAFS_BRONX_wage_HUD/JAFS_bronx_HUDwage_macro_2022.parquet")
df_JAFS_2022.head()

h_tract  year  access_jobs_total  access_low_total  \
0  36005000100  2022            3931761           2966555   
1  36005000200  2022           16710464          10133205   
2  36005000400  2022           24215019          14420841   
3  36005001600  2022           24170932          14548245   
4  36005001901  2022           14545039           7891939   

   access_median_total  access_high_total  access_low30  access_low50  \
0               606049              36106             0        371928   
1              5463641             118558             0       1662891   
2              8342360             135192             0       2382016   
3              8240186             110803             0       2407632   
4              5949956              69273             0       1592733   

   access_low60  access_low80  ...  JAFS_median_pc  JAFS_high_pc  \
0        239823       2354804  ...    30302.450000   1805.300000   
1       1619323       6850991  ...    60707.122222   1317.311111   
2       2292505       9746320  ...    44851.397849    726.838710   
3       2376675       9763938  ...     9537.252315    128.244213   
4       1606734       4692472  ...   145120.878049   1689.585366   

   JAFS_low30_pc  JAFS_low50_pc  JAFS_low60_pc  JAFS_low80_pc  JAFS_Office_pc  \
0            0.0   18596.400000   11991.150000  117740.200000    33177.150000   
1            0.0   18476.566667   17992.477778   76122.122222    65242.344444   
2            0.0   12806.537634   12325.295699   52399.569892    47986.844086   
3            0.0    2786.611111    2750.781250   11300.854167    10177.810185   
4            0.0   38847.146341   39188.634146  114450.536585   154207.487805   

   JAFS_Industrial_pc  JAFS_Institutional_pc  JAFS_LocalServices_pc  
0        74024.150000           42646.650000           30587.550000  
1        30229.055556           42675.155556           36469.044444  
2        19072.870968           30918.091398           25131.833333  
3         4067.339120            6721.201389            5537.392361  
4        40781.097561           66272.414634           78035.780488  

[5 rows x 27 columns]

In [ ]:
df_JAFS_2022.columns

Index(['h_tract', 'year', 'access_jobs_total', 'access_low_total',
       'access_median_total', 'access_high_total', 'access_low30',
       'access_low50', 'access_low60', 'access_low80', 'access_Office',
       'access_Industrial', 'access_Institutional', 'access_LocalServices',
       'resident_pop', 'JAFS_total_pc', 'JAFS_low_pc', 'JAFS_median_pc',
       'JAFS_high_pc', 'JAFS_low30_pc', 'JAFS_low50_pc', 'JAFS_low60_pc',
       'JAFS_low80_pc', 'JAFS_Office_pc', 'JAFS_Industrial_pc',
       'JAFS_Institutional_pc', 'JAFS_LocalServices_pc'],
      dtype='object')

In [ ]:
df_od_wac = pd.read_parquet("/content/drive/MyDrive/AI_DATA/JAFS_BRONX_wage_HUD/OD_WAC_detail_2022.parquet")
df_od_wac

h_tract      w_tract  year  S000  C000  CNS03  CNS04  CNS05  \
0        36005000200  36005000100  2022     1  1367      0      0      0   
1        36005000200  36005000100  2022     1  2158      0      0      0   
2        36005000200  36005000100  2022     1  3948      0      0      0   
3        36005000200  36005000100  2022     1  1367      0      0      0   
4        36005000200  36005000100  2022     1  2158      0      0      0   
...              ...          ...   ...   ...   ...    ...    ...    ...   
7445024  36005037300  42103950901  2022     1    15      0      0      0   
7445025  36005037300  42103950901  2022     1     9      0      2      0   
7445026  36005037300  42103950901  2022     1     1      0      0      0   
7445027  36005037300  42103950901  2022     1    24      0      0      0   
7445028  36005037300  42103950901  2022     1     9      0      4      0   

         CNS06  CNS07  ...  access_median_total  access_high_total  \
0            0      0  ...                    0                  0   
1            0      0  ...                    0                  0   
2            0      0  ...                    0                  0   
3            0      0  ...                    0                  0   
4            0      0  ...                    0                  0   
...        ...    ...  ...                  ...                ...   
7445024      0      0  ...                    0                  0   
7445025      0      0  ...                    1                  0   
7445026      0      0  ...                    0                  0   
7445027      0     12  ...                    0                  0   
7445028      0      5  ...                    0                  0   

         jobs_Office  jobs_Industrial  jobs_Institutional  jobs_LocalServices  \
0                  0                0                   0                   2   
1                  0                0                   0                   0   
2                  0                0                   0                   0   
3                  0                0                   0                   2   
4                  0                0                   0                   0   
...              ...              ...                 ...                 ...   
7445024            0                4                   0                  11   
7445025            1                8                   0                   0   
7445026            0                0                   1                   0   
7445027            0                0                   0                  24   
7445028            0                4                   0                   5   

         access_Office  access_Industrial  access_Institutional  \
0                    0                  0                     0   
1                    0                  0                     0   
2                    0                  0                     0   
3                    0                  0                     0   
4                    0                  0                     0   
...                ...                ...                   ...   
7445024              0                  4                     0   
7445025              1                  8                     0   
7445026              0                  0                     1   
7445027              0                  0                     0   
7445028              0                  4                     0   

         access_LocalServices  
0                           2  
1                           0  
2                           0  
3                           2  
4                           0  
...                       ...  
7445024                    11  
7445025                     0  
7445026                     0  
7445027                    24  
7445028                     5  

[7445029 rows x 34 columns]

Index(['h_tract', 'year', 'access_jobs_total', 'access_low_total',
       'access_median_total', 'access_high_total', 'access_low30',
       'access_low50', 'access_low60', 'access_low80', 'resident_pop',
       'JAFS_total_pc', 'JAFS_low_pc', 'JAFS_median_pc', 'JAFS_high_pc',
       'JAFS_low30_pc', 'JAFS_low50_pc', 'JAFS_low60_pc', 'JAFS_low80_pc'],
      dtype='object')

Index(['h_tract', 'year', 'access_jobs_total', 'access_low_total',
       'access_median_total', 'access_high_total', 'access_low30',
       'access_low50', 'access_low60', 'access_low80', 'access_Office',
       'access_Industrial', 'access_Institutional', 'access_LocalServices',
       'resident_pop', 'JAFS_total_pc', 'JAFS_low_pc', 'JAFS_median_pc',
       'JAFS_high_pc', 'JAFS_low30_pc', 'JAFS_low50_pc', 'JAFS_low60_pc',
       'JAFS_low80_pc', 'JAFS_Office_pc', 'JAFS_Industrial_pc',
       'JAFS_Institutional_pc', 'JAFS_LocalServices_pc'],
      dtype='object')